adapters/huaweiads/huaweiads.go

```diff
@@ -0,0 +1,1320 @@
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: "Bad Server Response",
+		}}
+	}
+
+	huaweiAdsResponseError := checkHuaweiAdsResponseRetcode(huaweiAdsResponse)
+	if huaweiAdsResponseError != nil {
+		return nil, []error{huaweiAdsResponseError}
+	}
+
+	var err error
+	bidderResponse, err = a.convertHuaweiAdsResp2BidderResp(&huaweiAdsResponse, openRTBRequest)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	return bidderResponse, nil
+}
+
+// Builder builds a new instance of the HuaweiAds adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+	bidder := &adapter{
+		endpoint: config.Endpoint,
+	}
+	return bidder, nil
+}
+
+// get request header, Authorization -> digest
+func getHeaders(huaweiAdsImpExt *openrtb_ext.ExtImpHuaweiAds, request *openrtb2.BidRequest, isAddAuthorization bool) http.Header {
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	if huaweiAdsImpExt == nil {
+		return headers
+	}
+
+	if isAddAuthorization {
+		headers.Add("Authorization", getDigestAuthorization(huaweiAdsImpExt))
+	}
+
+	if request.Device != nil && len(request.Device.UA) > 0 {
+		headers.Add("User-Agent", request.Device.UA)
+	}
+	return headers
+}
+
+// get body json for HuaweiAds request
+func getHuaweiAdsReqJson(request *huaweiAdsRequest, openRTBRequest *openrtb2.BidRequest, huaweiAdsImpExt *openrtb_ext.ExtImpHuaweiAds) error {
+	request.Version = HuaweiAdxApiVersion
+	var err error
+	if err = getHuaweiAdsReqAppInfo(request, openRTBRequest); err != nil {
+		return err
+	}
+	if err = getHuaweiAdsReqDeviceInfo(request, openRTBRequest, huaweiAdsImpExt); err != nil {
+		return err
+	}
+	getHuaweiAdsReqNetWorkInfo(request, openRTBRequest)
+	getHuaweiAdsReqRegsInfo(request, openRTBRequest)
+	getHuaweiAdsReqGeoInfo(request, openRTBRequest)
+	return nil
+}
+
+func getHuaweiAdsReqAdslot30(huaweiAdsImpExt *openrtb_ext.ExtImpHuaweiAds,
+	openRTBImp *openrtb2.Imp, openRTBRequest *openrtb2.BidRequest) (adslot30, error) {
+	adtypeLower := strings.ToLower(huaweiAdsImpExt.Adtype)
+	var adslot30 = adslot30{
+		Slotid: huaweiAdsImpExt.SlotId,
+		Adtype: convertAdtypeString2Integer(adtypeLower),
+		Test:   int32(openRTBRequest.Test),
+	}
+
+	if openRTBImp.Banner != nil {
+		if openRTBImp.Banner.W != nil && openRTBImp.Banner.H != nil {
+			adslot30.W = *openRTBImp.Banner.W
+			adslot30.H = *openRTBImp.Banner.H
+		}
+		if len(openRTBImp.Banner.Format) != 0 {
+			var formats = make([]format, 0, len(openRTBImp.Banner.Format))
+			for _, f := range openRTBImp.Banner.Format {
+				if f.H != 0 && f.W != 0 {
+					formats = append(formats, format{f.W, f.H})
+				}
+			}
+			adslot30.Format = formats
+		}
+	} else if openRTBImp.Native != nil {
+		if err := getNativeFormat(&adslot30, openRTBImp); err != nil {
```

Comment:
Please add a native test case which doesn't return an error. There doesn't seem to be one currently.

Reply:
already added nativeIncludeVide nativeSingleImage nativeThreeImage